src/Symfony/Component/DependencyInjection/Tests/Compiler/CheckDefinitionValidityPassTest.php

```diff
@@ -80,21 +80,46 @@ public function testValidTags()
         $container->register('b', 'class')->addTag('foo', ['bar' => null]);
         $container->register('c', 'class')->addTag('foo', ['bar' => 1]);
         $container->register('d', 'class')->addTag('foo', ['bar' => 1.1]);
+        $container->register('d', 'class')->addTag('foo', ['bar' => ['baz' => 'baz']]);
+        $container->register('e', 'class')->addTag('foo', ['deep' => ['foo' => ['bar' => 'baz']]]);
 
         $this->process($container);
 
         $this->addToAssertionCount(1);
     }
 
-    public function testInvalidTags()
+    /**
+     * @dataProvider provideInvalidTags
+     */
+    public function testInvalidTags(string $name, array $attributes, string $message)
     {
         $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage($message);
         $container = new ContainerBuilder();
-        $container->register('a', 'class')->addTag('foo', ['bar' => ['baz' => 'baz']]);
-
+        $container->register('a', 'class')->addTag($name, $attributes);
         $this->process($container);
     }
 
+    public function provideInvalidTags(): iterable
```

Comment:
should be static (will be asked by phpunit 10 for providers)

Reply:
OK, I didn't know about that 👍🏻